packages/changed-elements-react/src/api/ChangesTreeDataProvider.ts

```diff
@@ -474,24 +474,42 @@ export class ChangesTreeDataProvider implements ITreeDataProvider {
       updateFunc,
     );
 
-    let targetNodes = await this._getChangedModelOrSubjectNodes(
+    const targetNodes = await this._getChangedModelOrSubjectNodes(
       targetIModel,
       onlyDeleted,
       true,
       updateFunc,
     );
 
     const currentModelIds = new Set(currentNodes.map((node: TreeNodeItem) => node.id));
-    targetNodes = targetNodes.filter((entry: TreeNodeItem) => !currentModelIds.has(entry.id));
-
-    this._models = [...currentNodes, ...targetNodes];
+    const uniqueNodes = targetNodes.filter((entry: TreeNodeItem) => !currentModelIds.has(entry.id));
+    const mergedNodes = this._findAndMergeModelChanges(currentNodes, targetNodes);
+    this._models = [...mergedNodes, ...uniqueNodes];
 
     // Release, as it will be unused after the creation of models
     this._rootEntries = undefined;
 
     return this._models;
   }
 
+  private _findAndMergeModelChanges = (currentNodes: ReadonlyArray<TreeNodeItem>, targetNodes: ReadonlyArray<TreeNodeItem>) => {
+    const currentNodeMap = new Map<string, TreeNodeItem>();
+    currentNodes.forEach((node: TreeNodeItem) => {
+      currentNodeMap.set(node.id, node)
+    });
+    const overlappingNodes: ReadonlyArray<TreeNodeItem> = targetNodes.filter((entry: TreeNodeItem) => currentNodeMap.has(entry.id));
+    overlappingNodes.forEach((entry: TreeNodeItem) => {
+      const currentEntry = currentNodeMap.get(entry.id);
+      if (currentEntry?.extendedData?.modelChanges && entry.extendedData?.modelChanges) {
+        const keys = Object.keys(currentEntry?.extendedData?.modelChanges ?? "");
+        for (const key of keys) {
+          currentEntry.extendedData.modelChanges[key] = currentEntry.extendedData?.modelChanges[key] || entry.extendedData?.modelChanges[key];
+        }
+      }
+    });
+    return [...currentNodeMap.values()]
```

Comment:
Only relevant if this function is kept / needed at all:

May want to look at `Array.concat` if possible instead of using the spread operator (`...`), or find a diff solution. It can be slow with a lot of elements, and since these are the root elements of the models, they may be numerous enough to slow it down.

I've even seen spread operators crash altogether when dealing with big comparisons back in the day, so I'd like us to avoid them as much as possible.

Reply:
Can not use Array.concat because map.vaules returns an iterator so used Array.from instead